File: frontend/src/domain/entities/candidate.ts
export type StageResult = 'not_done' | 'passed' | 'rejected';
export type FinalStageResult = 'not_done' | 'offer' | 'rejected' | 'declined';
export type HireStatus = 'undecided' | 'hired' | 'offer_declined';

export interface Candidate {
  id: string;
  company_id: string;
  job_position_id: string;
  agent_id: string | null;
  name: string;
  resume_url: string | null;
  owner_user_id: string;
  note: string | null;
  stage_0_5_result: StageResult;
  stage_first_result: StageResult;
  stage_second_result: StageResult;
  stage_final_result: FinalStageResult;
  hire_status: HireStatus;
  mismatch_flag: boolean;
  stage_0_5_date: string | null;
  stage_first_date: string | null;
  stage_final_decision_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface CandidateWithRelations extends Candidate {
  company_name: string | null;
  job_position_name: string | null;
  agent_company_name: string | null;
  agent_contact_name: string | null;
  owner_user_name: string | null;
}

export interface CandidateCreate {
  company_id: string;
  job_position_id: string;
  agent_id?: string;
  name: string;
  resume_url?: string;
  owner_user_id: string;
  note?: string;
}

export interface CandidateUpdate {
  job_position_id?: string;
  agent_id?: string;
  name?: string;
  resume_url?: string;
  owner_user_id?: string;
  note?: string;
  stage_0_5_result?: StageResult;
  stage_first_result?: StageResult;
  stage_second_result?: StageResult;
  stage_final_result?: FinalStageResult;
  hire_status?: HireStatus;
  mismatch_flag?: boolean;
  stage_0_5_date?: string;
  stage_first_date?: string;
  stage_final_decision_date?: string;
}

export interface FunnelStats {
  total: number;
  stage_0_5_done: number;
  stage_0_5_passed: number;
  stage_first_done: number;
  stage_first_passed: number;
  stage_second_done: number;
  stage_second_passed: number;
  stage_final_done: number;
  stage_final_offer: number;
  hired: number;
  mismatch: number;
}

export interface MonthlyStats {
  active_candidates: number;
  hired: number;
  mismatch: number;
  stage_0_5_done: number;
  stage_0_5_passed: number;
}

export interface DashboardStats {
  active_candidates: number;
  pending_interviews: number;
  hired_this_month: number;
  mismatch_count: number;
  stage_0_5_pass_rate: number;
  stage_0_5_done_count: number;
  current_month: MonthlyStats;
  previous_month: MonthlyStats;
  active_trend_percent: number | null;
  hired_trend_percent: number | null;
}

File: api/app/application/dto/candidate.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel

from app.domain.entities.candidate import FinalStageResult, HireStatus, StageResult


class CandidateCreate(BaseModel):
    company_id: UUID
    job_position_id: UUID
    agent_id: UUID | None = None
    name: str
    resume_url: str | None = None
    owner_user_id: UUID
    note: str | None = None


class CandidateUpdate(BaseModel):
    job_position_id: UUID | None = None
    agent_id: UUID | None = None
    name: str | None = None
    resume_url: str | None = None
    owner_user_id: UUID | None = None
    note: str | None = None
    stage_0_5_result: StageResult | None = None
    stage_first_result: StageResult | None = None
    stage_second_result: StageResult | None = None
    stage_final_result: FinalStageResult | None = None
    hire_status: HireStatus | None = None
    mismatch_flag: bool | None = None
    stage_0_5_date: date | None = None
    stage_first_date: date | None = None
    stage_final_decision_date: date | None = None


class CandidateResponse(BaseModel):
    id: UUID
    company_id: UUID
    job_position_id: UUID
    agent_id: UUID | None
    name: str
    resume_url: str | None
    owner_user_id: UUID
    note: str | None
    stage_0_5_result: StageResult
    stage_first_result: StageResult
    stage_second_result: StageResult
    stage_final_result: FinalStageResult
    hire_status: HireStatus
    mismatch_flag: bool
    stage_0_5_date: date | None
    stage_first_date: date | None
    stage_final_decision_date: date | None
    created_at: datetime
    updated_at: datetime


class CandidateWithRelations(CandidateResponse):
    company_name: str | None = None
    job_position_name: str | None = None
    agent_company_name: str | None = None
    agent_contact_name: str | None = None
    owner_user_name: str | None = None


class CandidateFilter(BaseModel):
    company_id: UUID | None = None
    job_position_id: UUID | None = None
    agent_id: UUID | None = None
    owner_user_id: UUID | None = None
    stage_0_5_result: StageResult | None = None
    stage_final_result: FinalStageResult | None = None


class FunnelStats(BaseModel):
    total: int
    stage_0_5_done: int
    stage_0_5_passed: int
    stage_first_done: int
    stage_first_passed: int
    stage_second_done: int
    stage_second_passed: int
    stage_final_done: int
    stage_final_offer: int
    hired: int
    mismatch: int


class MonthlyStats(BaseModel):
    active_candidates: int
    hired: int
    mismatch: int
    stage_0_5_done: int
    stage_0_5_passed: int


class DashboardStats(BaseModel):
    active_candidates: int
    pending_interviews: int
    hired_this_month: int
    mismatch_count: int
    stage_0_5_pass_rate: float
    stage_0_5_done_count: int
    current_month: MonthlyStats
    previous_month: MonthlyStats
    active_trend_percent: float | None
    hired_trend_percent: float | None

